### PR TITLE
Faster multinomial implementation

### PR DESCRIPTION
**Pull request by cowanmeg**
New implementation removes the while loop. 
Following up from: https://github.com/pytorch/xla/pull/4848

### CODE REVIEW

**Review by wonjoolee95 — APPROVED** [review]
This is really cool. LGTM pending CI completion.

**Comment by wonjoolee95** [comment]
One thing -- can we try to enable the `multinomial` test we disabled in the previous PR, @cowanmeg?

File: torch_xla/csrc/xla_lower_util.cpp [hunk 1]
```diff
@@ -1154,48 +1154,18 @@ xla::XlaOp BuildMultinomial(xla::XlaOp input, int64_t num_samples,
...
+  cumprob = xla::BroadcastInDim(cumprob, broadcast_size, {dim});
+  rng = xla::BroadcastInDim(rng, broadcast_size, {input_shape.rank()});
 
-  std::vector<xla::XlaOp> results = ConsumeValue(xla::ForEachIndex(
-      sizes[dim], output_type, body_fn, {output, max_index, cumprob, rng},
-      "MultinomialMapping", output.builder()));
-  return results[0];
+  // Build comparison mask and sum along K dimension
+  auto output_type = xla::PrimitiveType::S64;
+  auto mask = BuildComparisonOp(at::aten::gt, rng, cumprob);
```
Comment (wonjoolee95) [inline torch_xla/csrc/xla_lower_util.cpp:1165]: This line seems to be responsible for the CI failure, related to shape differences between `rng` vs `cumprob`

Comment (cowanmeg) [reply torch_xla/csrc/xla_lower_util.cpp:1165]: Yeah, the BroadcastInDim calls didn't have proper arguments for 2D inputs.

**Comment by cowanmeg** [comment]
> One thing -- can we try to enable the `multinomial` test we disabled in the previous PR, @cowanmeg?

The failing test is: `test_multinomial_rng_state_advance`
Unfortunately, it still generates an error from long constant folding, so I'll keep it disabled.
I verified that the test passes when run isolated. 

```
 Constant folding an instruction is taking > 1s:

  reduce-window.70 (displaying the full instruction incurs a runtime overhead. Raise your logging level to 4 or above).

This isn't necessarily a bug; constant-folding is inherently a trade-off between compilation time and speed at runtime.  XLA has some guards that attempt to keep constant folding from taking too long, but fundamentally you'll always be able to come up with an input program that takes a long time.
```

**Review by wonjoolee95 — APPROVED** [review]
LGTM, thanks!

**Comment by miladm** [comment]
Thanks @cowanmeg. Wondering if verified the perf latency gain of this op with/without the fix in python unit test script?

**Pull request merged**